pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,105 @@
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
+    def __init__(self,
```

Comment:
@frtizo - As discussed, I have removed the dependency on Dice, and moved the log prob computation to its own class to make it easy to test (test cases are added in `infer/mcmc/test_valid_models.py`). I think this should be self-explanatory (and very similar to the Dice logic), but let me know in case I may have overlooked some corner cases.

Reply:
Also, this should resolve the bugs in the previous implementation. I realized that the logic of only looking at the enumerated dims when computing the upstream log probs was wrong.